jsettlers.logic/src/main/java/jsettlers/ai/highlevel/WhatToDoAi.java

```diff
@@ -361,15 +411,70 @@ private void addBuildingCountToBuildingPlan(EBuildingType buildingType, Map<EBui
 		playerBuildingPlan.put(buildingType, playerBuildingPlan.get(buildingType) + 1);
 	}
 
-	private boolean buildTower() {
-		if (aiStatistics.getTotalNumberOfBuildingTypeForPlayer(STONECUTTER, playerId) >= 1
-				&& aiStatistics.getNumberOfNotOccupiedMilitaryBuildings(playerId) == 0) {
-			destroyHinterlandMilitaryBuildings();
-			return construct(TOWER);
+	private void commandPioneers() {
+		if (aiStatistics.getBorderOf(playerId).size() == 0 || aiStatistics.getEnemiesInTownOf(playerId).size() > 0) {
+			releaseAllPioneers();
+		} else if (aiStatistics.getNumberOfTotalBuildingsForPlayer(playerId) >= 4) {
+			sendOutPioneers();
 		}
-		return false;
 	}
 
+	private void releaseAllPioneers() {
+		broadenerPioneers.clear();
+		resourcePioneers.clear();
+		List<ShortPoint2D> pioneers = aiStatistics.getMovablePositionsByTypeForPlayer(EMovableType.PIONEER, playerId);
+		if (!pioneers.isEmpty()) {
+			List<Integer> pioneerIds = new ArrayList<>(pioneers.size());
+			for (ShortPoint2D pioneerPosition : pioneers) {
+				pioneerIds.add(mainGrid.getMovableGrid().getMovableAt(pioneerPosition.x, pioneerPosition.y).getID());
+			}
+			taskScheduler.scheduleTask(new ConvertGuiTask(playerId, pioneerIds, EMovableType.BEARER));
+			//pioneers which can not be converted shall walk into player's land to be converted the next tic
+			taskScheduler.scheduleTask(new MoveToGuiTask(playerId, aiStatistics.getPositionOfPartition(playerId), pioneerIds));
+		}
+	}
+
+	private void sendOutPioneers() {
+		resourcePioneers.removeDeadPioneers();
+		broadenerPioneers.removeDeadPioneers();
+
+		if (!resourcePioneers.isFull()) {
```

Comment:
If I would do this the code would become more complex because of the asynchronous execution of the AI orders.

The Ids of the bearers must be remembered to not double convert a bearer into a pioneer for both groups.

I think the algorithm is better than using no pioneers at all. So we should merge the PR into the codebase and cover optimisations later following the principle to do the Must Haves first and the Nice to Haves later.